enumfields/fields.py

```diff
@@ -31,7 +32,38 @@ def __set__(self, obj, value):
         obj.__dict__[self.field.name] = self.field.to_python(value)
 
 
-class EnumFieldMixin:
+class EnumFieldValidationMixin:
+
+    def _raise_error_if_next_value_not_allowed(self, value, model_instance):
+        if model_instance.pk is not None:
+            previous_values = [choice.value for choice in self.enum.get_allowed_previous_choices(value.name)]
+
+            # search whether in db exists record within all posible previous states and the current one
+            if not self.model.objects.filter(
+                Q(**{'{}__in'.format(self.name): previous_values + [value.value]}) | Q(**{self.name: None}),
+                pk=model_instance.pk,
+            ).exists():
+                raise ValidationError(
+                    ugettext(
+                        'Transition from current choice to {} choice is not allowed').format(
+                        value.name)
+                )
+
+    def _raise_error_if_initial_value_not_allowed(self, value, model_instance):
+        if model_instance.id is None:
```

Comment:
this is not right solution. Because not all models has auto generated id. For this models you set id before its save, but instance is still new. You should use ._state.adding

https://docs.djangoproject.com/en/2.2/ref/models/instances/#customizing-model-loading

Reply:
At least, you should use `pk` instead of `id` (as you have above). The primary key might not be `id`.

Reply:
no this is not still right solution :) it is the same as instance.id

Reply:
No all the time, `pk` always points to model's primary key. If you decide your model's primary key to be `magic_id`, `instance.pk` will still work, whereas `instance.id` will stop working. So even if model does not have auto generated id, but has primary key, `pk` will work imho.
https://stackoverflow.com/questions/2165865/django-queries-id-vs-pk

Reply:
good ideas from both, I am using `._state.adding` now

Reply:
@rubickcz I think you do not understand the problem. self.pk or self.id cannot be used to check if instance is adding or changing, because for model with not auto generated FK the condition `self.pk is None` will not work.

Reply:
Ok ok, you are right.